plugins/funind/indfun_common.ml

```diff
@@ -378,7 +377,72 @@ let () = declare_bool_option function_debug_sig
 
 let do_observe () = !function_debug
 
+let observe strm =
+  if do_observe ()
+  then Feedback.msg_debug strm
+  else ()
+
+let debug_queue = Stack.create ()
+
+let print_debug_queue b e =
+  if not (Stack.is_empty debug_queue)
+  then
+    let lmsg,goal = Stack.pop debug_queue in
+    (if b then
+       Feedback.msg_debug (hov 1 (lmsg ++ (str " raised exception " ++ CErrors.print e) ++ str " on goal" ++ fnl() ++ goal))
+     else
+       Feedback.msg_debug (hov 1 (str " from " ++ lmsg ++ str " on goal"++fnl() ++ goal))
+       (* print_debug_queue false e; *)
+    )
 
+let do_observe_tac s tac g =
+  let goal = Printer.pr_goal g in
+  let s = s (pf_env g) (project g) in
+  let lmsg = (str "observation : ") ++ s in
+  Stack.push (lmsg,goal) debug_queue;
+  try
+    let v = tac g in
+    ignore(Stack.pop debug_queue);
+    v
+  with reraise ->
+    let reraise = CErrors.push reraise in
+    if not (Stack.is_empty debug_queue)
+    then print_debug_queue true (fst reraise);
+    Util.iraise reraise
+
+let observe_tac s tac g =
+  if do_observe ()
+  then do_observe_tac s tac g
+  else tac g
+
+module New = struct
+
+let do_observe_tac ~header s tac =
```

Comment:
The implementation below is incorrect. Binding `tac` cannot raise an exception, thus the try with block is never going through the catch branch. If anything, you should try to use `tclZERO` and `tclOR` to observe tactic failure. Nonetheless, this is quite horrible debug code which I am not sure it is worthwhile to port. I wouldn't lose sleep over its removal.

Reply:
Indeed it makes no sense; will fix adding some printf using the non-logical monad.

I'd love to remove that layer which pollutes code a big deal but on the other hand I'm unsure about that as it could render the plugin undebuggable and I am not familiar with the existing facilities to debug tactics. I guess as the code does improve we will be able to remove it some day.

Reply:
Let's see now...

Reply:
Seems like CI worked.

Reply:
This is most likely still broken, but I prefer to merge this PR as-is to not delay further cleanups. The original code was probably already quite broken and it is for debugging purposes, so I don't care that much.

Reply:
Thanks @ppedrot , further ideas on how to proceed are indeed welcome.